File: structures-js/structures-api/src/api/idl/decorators/DiscriminatorDecorator.ts
import {C3Decorator} from '@kinotic/continuum-idl'

/**
 * Specifies the discriminator field for polymorphic entities.
 */
export class DiscriminatorDecorator extends C3Decorator {
    public propertyName?: string
    constructor() {
        super()
        this.type = 'Discriminator'
    }

    public withPropertyName(propertyName: string): DiscriminatorDecorator {
        this.propertyName = propertyName
        return this
    }
}

File: structures-api/src/main/java/org/kinotic/structures/api/domain/idl/decorators/DiscriminatorDecorator.java
package org.kinotic.structures.api.domain.idl.decorators;

import com.fasterxml.jackson.annotation.JsonIgnore;
import lombok.Getter;
import lombok.Setter;
import lombok.experimental.Accessors;
import org.kinotic.continuum.idl.api.schema.decorators.C3Decorator;
import org.kinotic.continuum.idl.api.schema.decorators.DecoratorTarget;

import java.util.List;

/**
 * Created by Navíd Mitchell 🤪on 6/16/23.
 */
@Getter
@Setter
@Accessors(chain = true)
public final class DiscriminatorDecorator extends C3Decorator {

    @JsonIgnore
    public static final String type = "Discriminator";

    private String propertyName = null;

    public DiscriminatorDecorator() {
        this.targets = List.of(DecoratorTarget.FIELD);
    }
}
